From 2e309690fbcc0a9c71303f1f97352e051e276e51 Mon Sep 17 00:00:00 2001
From: Bijoy Jose <bijoy.a.jose@intel.com>
Date: Thu, 3 Jan 2013 15:34:45 -0800
Subject: Dalvik: Scheduler needs DIV support

BZ: 75820

The Atom instruction scheduler needs to know instruction latency
and port usage for DIV instruction. Otherwise, it may come up
with a non-optimal schedule.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I96855fdb4d1042dc7b2158c707c5188d67c97e3f
Orig-MCG-Change-Id: I483e71b72eedbbc01cf0ba60bcd89c187721bb36
Signed-off-by: Razvan A Lupusoru <razvan.a.lupusoru@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/Schedule.cpp |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/vm/compiler/codegen/x86/Schedule.cpp b/vm/compiler/codegen/x86/Schedule.cpp
index 4edfcbd..16b12c8 100644
--- a/vm/compiler/codegen/x86/Schedule.cpp
+++ b/vm/compiler/codegen/x86/Schedule.cpp
@@ -102,7 +102,7 @@ struct MachineModelEntry {
 //! \def INVN
 //! \brief This is an abbreviation of invalid node latency and is used for
 //! readability reasons.
-#define INVN -1
+#define INVN 0
 
 //! \def REG_NOT_USED
 //! \brief This is an abbreviation for register not used and is used for
@@ -226,7 +226,8 @@ MachineModelEntry atomMachineModel[Mnemonic_Count*6] = {
     {INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //FSCALE
 
     {INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //XCHG
-    {INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //DIV
+    // There is no way to differentiate operand sizes in this table, so just assume 32-bit
+    {INVP,INVN},{BOTH_PORTS,57},{BOTH_PORTS,57},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //DIV
     {INVP,INVN},{BOTH_PORTS,57},{BOTH_PORTS,57},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //IDIV
     {INVP,INVN},{BOTH_PORTS,6},{BOTH_PORTS,7},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //MUL
     // This table does not support IMUL with single reg or mem operand
-- 
1.7.4.1

